### PR TITLE
Add query queued and running time limits to resource groups

### PR DESCRIPTION
**Pull request by elonazoulay**
Added 2 new properties to the resource groups: queuedTimeLimit and runningTimeLimit.

### CODE REVIEW

**Comment by elonazoulay** [comment]
Resolves #6468

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 1]
```diff
@@ -630,6 +675,34 @@ protected boolean internalStartNext()
...
+    protected void enforceTimeLimits()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to enforce time limits");
+        synchronized (root) {
+            for (InternalResourceGroup group : subGroups.values()) {
+                group.enforceTimeLimits();
+            }
+            long runningTimeoutMillis = runningTimeout.toMillis();
+            long queuedTimeoutMillis = queuedTimeout.toMillis();
+            // Copy of activeQueries since queryFinished gets called by cancelQuery which modifies activeQueries
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Copy active queries to prevent ConcurrentModificationException

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 2]
```diff
@@ -630,6 +675,34 @@ protected boolean internalStartNext()
...
+            for (QueryExecution query : ImmutableList.copyOf(activeQueries)) {
+                if (runningQueries.contains(query)) {
+                    if (Optional.ofNullable(query.getQueryInfo().getQueryStats().getExecutionTime())
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Pull this out into a variable, and have the first if AND with it - no need for nesting.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 3]
```diff
@@ -630,6 +675,34 @@ protected boolean internalStartNext()
...
     }
 
+    protected void enforceTimeLimits()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to enforce time limits");
+        synchronized (root) {
+            for (InternalResourceGroup group : subGroups.values()) {
+                group.enforceTimeLimits();
+            }
+            long runningTimeoutMillis = runningTimeout.toMillis();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Durations are comparable, no need to convert like this.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 4]
```diff
@@ -630,6 +675,34 @@ protected boolean internalStartNext()
...
+            long queuedTimeoutMillis = queuedTimeout.toMillis();
+            // Copy of activeQueries since queryFinished gets called by cancelQuery which modifies activeQueries
+            for (QueryExecution query : ImmutableList.copyOf(activeQueries)) {
+                if (runningQueries.contains(query)) {
+                    if (Optional.ofNullable(query.getQueryInfo().getQueryStats().getExecutionTime())
+                            .orElse(new Duration(0, NANOSECONDS)).toMillis() > runningTimeoutMillis) {
+                        cachedMemoryUsageBytes -= query.getTotalMemoryReservation();
+                        query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "query exceeded resource group runtime limit"));
+                    }
+                } // Queued time may be null when a query is queued, subtract execution time from elapsed time instead
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Move to next line.
I think queued time is *always* null when the query is queued.
Just pull out the 'queued' time into a variable - make your conditionals simpler.

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java [hunk 5]
```diff
@@ -177,13 +177,15 @@ protected void configureGroup(ResourceGroup group, ResourceGroupSpec match)
...
+        match.getQueuedTimeout().ifPresent(timeout -> group.setQueuedTimeLimit(timeout));
+        match.getRunningTimeout().ifPresent(timeout -> group.setRunningTimeLimit(timeout));
         if (match.getSchedulingPolicy().isPresent()) {
             group.setSchedulingPolicy(match.getSchedulingPolicy().get());
         }
         if (match.getSchedulingWeight().isPresent()) {
             group.setSchedulingWeight(match.getSchedulingWeight().get());
         }
-        if (match.getJmxExport().isPresent()) {
+        if (match.getJmxExport().isPresent() && match.getJmxExport().get() != group.getJmxExport()) {
```
Comment (raghavsethi) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java]: What is this about?

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java [hunk 6]
```diff
@@ -177,13 +177,15 @@ protected void configureGroup(ResourceGroup group, ResourceGroupSpec match)
         }
         group.setMaxQueuedQueries(match.getMaxQueued());
         group.setMaxRunningQueries(match.getMaxRunning());
+        match.getQueuedTimeout().ifPresent(timeout -> group.setQueuedTimeLimit(timeout));
```
Comment (raghavsethi) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java]: Use the if(isPresent) style to be consistent with the surrounding code

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.IDBI;
+import org.skife.jdbi.v2.exceptions.DBIException;
+
+import java.lang.reflect.InvocationTargetException;
+
+import static com.google.common.reflect.Reflection.newProxy;
+import static java.util.Objects.requireNonNull;
+
+public class DatabaseUtil
```
Comment (raghavsethi) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java]: Wait what's going on here - this seems like a totally separate change and should be a different commit

Comment (elonazoulay) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java]: Will put in a separate commit, let me know if you want it to be in a separate pr as well.

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java (hunk 5 above)
Comment (elonazoulay) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java]: Without that an exception is thrown if the previous and new values are the same. I'll add a comment.

**Comment by elonazoulay** [comment]
Looks like I do not have permission to assign back. Let me know if everything looks ok. Thanks!

**Comment by raghavsethi** [comment]
Maven checks are failing

**Comment by elonazoulay** [comment]
Will rebase and push.

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+{
+    // Partial copy from raptor: com.facebook.presto.raptor.util.DatabaseUtil
```
Comment (raghavsethi) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java]: Put this comment next to the onDemandDao method

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java (hunk 5 above)
Comment (raghavsethi) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/AbstractResourceConfigurationManager.java]: I understand - my question is whether this should be a separate commit or not?

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DbResourceGroupConfigurationManager.java [hunk 9]
```diff
@@ -135,25 +137,30 @@ public void configure(ResourceGroup group, SelectionContext context)
 
     private synchronized void load()
     {
-        Map.Entry<ManagerSpec, Map<ResourceGroupIdTemplate, ResourceGroupSpec>> specsFromDb = buildSpecsFromDb();
-        ManagerSpec managerSpec = specsFromDb.getKey();
-        Map<ResourceGroupIdTemplate, ResourceGroupSpec> resourceGroupSpecs = specsFromDb.getValue();
-        Set<ResourceGroupIdTemplate> changedSpecs = new HashSet<>();
-        Set<ResourceGroupIdTemplate> deletedSpecs = Sets.difference(this.resourceGroupSpecs.keySet(), resourceGroupSpecs.keySet());
+        try {
```
Comment (raghavsethi) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DbResourceGroupConfigurationManager.java]: Not clear to me what this change has to do with queued and running time limits?

File: presto-tests/src/test/java/com/facebook/presto/execution/resourceGroups/db/TestQueues.java [hunk 10]
```diff
@@ -119,7 +119,7 @@ public void testBasic()
         }
     }
 
-    @Test(timeOut = 240_000)
+    @Test(timeOut = 60_000)
```
Comment (raghavsethi) [inline presto-tests/src/test/java/com/facebook/presto/execution/resourceGroups/db/TestQueues.java]: Could you please move unrelated changes to a separate commit?

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+    private DatabaseUtil() {}
+
+    public static <T> T onDemandDao(IDBI dbi, Class<T> daoType)
```
Comment (electrum) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DatabaseUtil.java]: This method is not needed. You can just use dbi.onDemand.

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DbResourceGroupConfigurationManager.java (hunk 9 above)
Comment (elonazoulay) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/DbResourceGroupConfigurationManager.java]: I will save this for another pr related to stale connections.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 12]
```diff
@@ -110,9 +114,15 @@
...
     @GuardedBy("root")
+    private LinkedHashSet<QueryExecution> activeQueries = new LinkedHashSet<>();
+    @GuardedBy("root")
     private SchedulingPolicy schedulingPolicy = FAIR;
     @GuardedBy("root")
     private boolean jmxExport;
+    @GuardedBy("root")
+    private Duration queuedTimeLimit = new Duration(Long.MAX_VALUE, MILLISECONDS);
+    @GuardedBy("root")
+    private Duration runningTimeLimit = new Duration(Long.MAX_VALUE, MILLISECONDS);;
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Fix double semicolon. It's weird checkstyle didn't catch that

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 13]
```diff
@@ -110,9 +114,15 @@
     @GuardedBy("root")
     private final Set<QueryExecution> runningQueries = new HashSet<>();
     @GuardedBy("root")
+    private LinkedHashSet<QueryExecution> activeQueries = new LinkedHashSet<>();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: I thought about this more last night, and I'm not sure there's value in having another data structure that is basically just queued+running. Let's not add another data structure that can get out of sync/leak.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 14]
```diff
@@ -656,6 +721,31 @@ protected boolean internalStartNext()
...
     }
 
+    protected void enforceTimeLimits()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to enforce time limits");
+        synchronized (root) {
+            for (InternalResourceGroup group : subGroups.values()) {
+                group.enforceTimeLimits();
+            }
+            // Copy active queries to prevent ConcurrentModificationException
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Remove comment

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 15]
```diff
@@ -375,41 +382,15 @@ public void setSchedulingPolicy(SchedulingPolicy policy)
             if (parent.isPresent() && parent.get().schedulingPolicy == QUERY_PRIORITY) {
                 checkArgument(policy == QUERY_PRIORITY, "Parent of %s uses query priority scheduling, so %s must also", id, id);
             }
-
-            // Switch to the appropriate queue implementation to implement the desired policy
-            UpdateablePriorityQueue<InternalResourceGroup> queue;
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Why was this refactor made after the code was already accepted? Please revert, and put in a separate commit if you want.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 14 above)
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: I reworded the comment to specify that queued and running queries are copied.

**Comment by elonazoulay** [comment]
I reverted the commit and added a new commit with explanation of the refactor (to remove activeQueries). Let me know if this looks ok.

**Comment by elonazoulay** [comment]
Found and cleaned up code that used running and queued "Timeout" instead of "TimeLimit" - added a new commit. Let me know if this should be included, I can revert this as well.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 16]
```diff
@@ -656,6 +700,27 @@ protected boolean internalStartNext()
...
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to enforce time limits");
+        synchronized (root) {
+            for (InternalResourceGroup group : subGroups.values()) {
+                group.enforceTimeLimits();
+            }
+            // Copy active queries to prevent ConcurrentModificationException
+            for (QueryExecution query : ImmutableList.copyOf(activeQueries)) {
+                Optional<Duration> runningTime = Optional.ofNullable(query.getQueryInfo().getQueryStats().getExecutionTime());
+                Optional<Duration> elapsedTime = Optional.ofNullable(query.getQueryInfo().getQueryStats().getElapsedTime());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Don't need an optional here, just check for null in the ifs.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/IndexedPriorityQueue.java [hunk 17]
```diff
@@ -111,6 +114,12 @@ public boolean isEmpty()
         return queue.isEmpty();
     }
 
+    @Override
+    public ImmutableList<E> toList()
+    {
+        return queue.stream().map(Entry::getValue).collect(ImmutableList.toImmutableList());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/IndexedPriorityQueue.java]: Static import

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/UpdateablePriorityQueue.java [hunk 18]
```diff
@@ -28,4 +30,6 @@
     int size();
 
     boolean isEmpty();
+
+    ImmutableList<E> toList();
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/UpdateablePriorityQueue.java]: `List<E> toList`

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FifoQueue.java [hunk 19]
```diff
@@ -74,4 +74,10 @@ public boolean isEmpty()
     {
         return delegate.isEmpty();
     }
+
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FifoQueue.java]: Change commit message to:

Change UpdateablePriorityQueue to implement Iterable

Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FifoQueue.java]: Updated, thanks!

**Comment by raghavsethi** [comment]
Merged, thanks for your contribution!
